README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# predictr <a href='https://kiernann.com/predictr'><img src='man/figures/logo.png' align="right" height="139" /></a>

<!-- badges: start -->

[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-maturing-blue.svg)](https://www.tidyverse.org/lifecycle/#experimental)
[![CRAN
status](https://www.r-pkg.org/badges/version/predictr)](https://CRAN.R-project.org/package=predictr)
[![Travis build
status](https://travis-ci.org/kiernann/predictr.svg?branch=master)](https://travis-ci.org/kiernann/predictr)
[![Codecov test
coverage](https://img.shields.io/codecov/c/github/kiernann/predictr/master.svg)](https://codecov.io/gh/kiernann/predictr?branch=master)
<!-- badges: end -->

The goal of predictr is to access the
[PredictIt](https://www.predictit.org/) API and determine the price of
binary futures contracts trading on the prediction market. Data is
returned as tidy tibbles, with unique rows for contract prices at a
given point in time.

## Markets

[Prediction markets](https://en.wikipedia.org/wiki/Prediction_market)
demonstrate the [efficient market
hypothesis](https://en.wikipedia.org/wiki/Efficient-market_hypothesis),
which states that market prices reflect *all* available information.
Self-interested traders buy and sell contracts tied to a binary outcome.
These contracts expire for $1 or $0, depending on the outcome.

The more likely an outcome is, the higher the price and the lower the
payout. As the probability of an outcome changes, traders buy or sell at
new prices and the equilibrium price updates to reflect the new
probability.

## Installation

You can install the development version from
[GitHub](https://github.com/kiernann/predictr) with:

``` r
# install.packages("devtools")
devtools::install_github("kiernann/predictr")
```

## Usage

The `open_markets()` function lists all active markets.

``` r
library(predictr)
library(tidyverse)
all_markets <- open_markets(split = FALSE)
```

Use `market_price()` to get the *current* price of a market’s contracts.

``` r
# Electoral College margin of victory?
market_price(6653)[-3]
#> # A tibble: 16 x 7
#>    time                  mid   cid contract           last close end                
#>    <dttm>              <int> <int> <fct>             <dbl> <dbl> <dttm>             
#>  1 2020-07-16 13:10:47  6653 22320 GOP by 280+        0.02  0.03 NA                 
#>  2 2020-07-16 13:10:47  6653 22347 GOP by 210 - 279   0.02  0.02 NA                 
#>  3 2020-07-16 13:10:47  6653 22345 GOP by 150 - 209   0.04  0.04 NA                 
#>  4 2020-07-16 13:10:47  6653 22325 GOP by 100 - 149   0.05  0.05 NA                 
#>  5 2020-07-16 13:10:47  6653 22319 GOP by 60 - 99     0.06  0.07 NA                 
#>  6 2020-07-16 13:10:47  6653 22317 GOP by 30 -59      0.06  0.06 NA                 
#>  7 2020-07-16 13:10:47  6653 22316 GOP by 10 - 29     0.06  0.06 NA                 
#>  8 2020-07-16 13:10:47  6653 22315 GOP by 0 - 9       0.06  0.06 NA                 
#>  9 2020-07-16 13:10:47  6653 22321 Dems by 1 - 9      0.03  0.03 NA                 
#> 10 2020-07-16 13:10:47  6653 22322 Dems by 10 - 29    0.04  0.04 NA                 
#> 11 2020-07-16 13:10:47  6653 22323 Dems by 30 -59     0.07  0.06 NA                 
#> 12 2020-07-16 13:10:47  6653 22318 Dems by 60 - 99    0.09  0.09 NA                 
#> 13 2020-07-16 13:10:47  6653 22326 Dems by 100 - 149  0.12  0.12 NA                 
#> 14 2020-07-16 13:10:47  6653 22346 Dems by 150 - 209  0.15  0.14 NA                 
#> 15 2020-07-16 13:10:47  6653 22348 Dems by 210 - 279  0.14  0.13 NA                 
#> 16 2020-07-16 13:10:47  6653 22324 Dems by 280+       0.17  0.16 NA
```

<img src="man/figures/README-plot-price-1.png" width="100%" />

Using `market_history()` can return 24 hours or 90 days of price
history.

<img src="man/figures/README-plot-history-1.png" width="100%" />

<!-- refs: start -->

<!-- refs: end -->


NEWS.md
# predictr 0.2.1

* Add back `contract_convert()` using only base R functions.

# predictr 0.2.0

* All functions now use `httr::RETRY()` instead of `GET()` on its own.
* Removed `contract_convert()` (and moved to [k5] package).
* All returned data now uses the default column names and structure.

[k5]: https://github.com/kiernann/k5/

# predictr 0.1.0

* All functions have been rewritten to use `httr::GET()` instead of attempting
to read the JSON or CSV files directly.
* The `market_history()` function now only makes a single request for the CSV
data, forgoing off the call to the API for the contract IDs and market question.
* `market_plot()` has been removed to reduce dependencies.
* Contract values are sorted by the `displayOrder` API value.
* All `contract_convert()` has been removed.

# predictr 0.0.2

* Create `market_plot()` using ggplot2.
* Order `market_price()` by contract factor.
* Set market names as character strings.
* Convert binary contracts to logical.

# predictr 0.0.1

* Create `market_history()`.
* Remove `dplyr` complexity to avoid global variables.
* Improve interval contract factorization.

# predictr 0.0.0.9001

* Create `market_price()` for a single market call.
* Implement factorization of interval contract names.
* Allow for `dplyr::group_split()` in `open_markets()`.

# predictr 0.0.0.9000

* Create `open_markets()` to list all in tidy data frane.
* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
library(testthat)
library(predictr)

test_check("predictr")


tests/testthat/test-market-plot.R
library(testthat)
library(predictr)

test_that("market history plot can be made", {
  x <- market_history(3633)
  p <- market_plot(x)
  expect_s3_class(p, "ggplot")
})

test_that("market history plot can be made", {
  x <- market_price(6405)
  p <- market_plot(x)
  expect_s3_class(p, "ggplot")
})


tests/testthat/test-market-price.R
library(testthat)
library(predictr)

test_that("discrete market price returns", {
  dat <- market_price(id = 3633)
  expect_length(dat, 16)
  expect_s3_class(dat, "data.frame")
  expect_s3_class(dat$timeStamp, "POSIXct")
  expect_s3_class(dat$dateEnd, "POSIXct")
})


R/interval.R
#' Convert contract names to factor intervals
#'
#' Can perform one of three **rough** conversions:
#' 1. For interval contracts (e.g., "220 - 229", "9% or more", etc.), convert
#' the character strings to proper interval notation.
#' 2. For contracts with multiple discrete outcomes (e.g., Candidate names),
#' convert the character vector to simple factors.
#' 3. For markets with a single binary question (e.g., "Will the Democrats have
#' a brokered convention in 2020?"), contracts returned are always "Yes" which
#' is converted to `TRUE`.
#'
#' @param x A character vector of contract names.
#' @return A interval factor, unique factor, or logical vector.
#' @examples
#' contract_convert(c("4 or fewer", "8 or more", "5", "6", "7"))
#' contract_convert(c("3%-", "12%+", "9% - 12%", "3% - 6%", "6% - 9%"))
#' contract_convert(c("219-", "220 to 229", "240+",  "230 to 239"))
#' contract_convert(c("Sanders", "Biden", "Warren", "Buttigieg"))
#' contract_convert(c("Brokered convention?", "Brokered convention?"))
#' @export
contract_convert <- function(x) {
  # regex for any number with separators
  rx <- "(\\d{1,3}(?:,\\d{3})*(?:\\.\\d+)?)"
  # check for percentafes before removing
  p <- all(grepl("%", x)) # check for percentages
  if (length(unique(x)) == 1) {
    # one value matches market name
    return(TRUE)
  } else if (all(!grepl("\\d", x))) {
    # return characters as factor
    return(factor(x, ordered = FALSE))
  }
  # replace words with symbols
  x <- gsub("\\sor\\s(fewer|lower)", "-", x)
  x <- gsub("\\sor\\s(more|higher)", "+", x)
  x <- gsub("\\sto\\s", " - ", x)
  # remove all non symbols/numbers
  x <- gsub("[^\\+\\d\\s\\.-]", "", x, perl = TRUE)
  # remove .0 decimal before non-number
  x <- gsub("(?<=\\d)(\\.0)(?=\\D)", "", x, perl = TRUE)
  # if any value is a single number
  if (any(grepl(sprintf("^%s$", rx), x))) {
    # place number before negative at the end
    x <- gsub(paste0(rx, "(?:-$)"), "[0,\\1]", x)
    # place number before positive at the start to inf
    x <- gsub(paste0(rx, "(?:\\+$)"), "[\\1, Inf)", x)
    # if only numbers put in own interval
    x <- gsub(sprintf("^%s$", rx), "[\\1,\\1]", x)
  } else if (p) { # if all percentages
    # place percentage inbetween 1 and 0
    x <- gsub(paste0(rx, "(?:-$)"), "[0,\\1)", x)
    x <- gsub(paste0(rx, "(?:\\+$)"), "[\\1,100]", x)
    x <- gsub(paste(rx, "-", rx), "[\\1,\\2)", x)
  } else if (any(grepl(paste(rx, "-", rx), x))) {
    x <- gsub(paste0(rx, "(?:-$)"), "[0,\\1]", x)
    x <- gsub(paste(rx, "(?:-|to)", rx), "[\\1,\\2]", x)
    x <- gsub(paste0(rx, "(?:\\+$)"), "[\\1, Inf)", x)
  }
  z <- unique(x)
  a <- regmatches(x, regexpr("(\\d{1,3}(?:\\.\\d+)?)|Inf", x))
  a <- vapply(a, FUN = function(i) sum(as.numeric(i)), FUN.VALUE = double(1))
  a <- match(a, sort(a))
  factor(x, levels = z[order(sort(z)[a])], ordered = TRUE)
}


R/market-price.R
#' Current market prices
#'
#' @param id The integer market ID.
#' @return A data frame of market contracts prices.
#' @examples
#' market_price(id = 6653)
#' @importFrom httr GET content accept_json
#' @export
market_price <- function(id) {
  resp <- httr::RETRY(
    verb = "GET",
    url = paste0("https://www.predictit.org/api/marketdata/markets/", id),
    httr::accept_json()
  )
  dat <- httr::content(
    x = resp,
    as = "parsed",
    type = "application/json",
    encoding = "ASCII",
    flatten = TRUE,
    simplifyDataFrame = TRUE
  )

  names(dat$contracts)[1] <- "contract"
  dat$contracts$dateEnd <- api_time(dat$contracts$dateEnd)
  cont <- cbind(
    id = id,
    market = dat$name,
    shortMarket = dat$shortName,
    timeStamp = api_time(dat$timeStamp),
    dat$contracts
  )
  as_tibble(cont[, -c(7)])
}


R/market-history.R
#' Market price history
#'
#' Download a data frame of contract prices over time, either hourly or daily.
#' This chart data does not include a market name or numeric contract IDs, which
#' can be obtained from the list of all [open_markets()].
#'
#' @param id Market ID number.
#' @param timespan One of "24h", "7d" or "90d". Either hourly or daily.
#' @return A data frame of market contract prices over time.
#' @examples
#' market_history(id = 6653, timespan = "7d")
#' @importFrom httr RETRY content
#' @importFrom utils read.csv
#' @export
market_history <- function(id, timespan = c("24h", "7d", "90d")) {
  resp <- httr::RETRY(
    verb = "GET",
    url = "https://www.predictit.org/Resource/DownloadMarketChartData",
    query = list(marketid = id, timespan = match.arg(timespan))
  )
  dat <- httr::content(
    x = resp,
    as = "text",
    type = "text/csv",
    encoding = "ASCII"
  )
  con <- textConnection(dat)
  on.exit(con)
  dat <- utils::read.csv(con)
  dat$Date <- as.POSIXct(dat$Date, tz = "EDT", format = "%m/%d/%Y %H:%M:%S")
  dat[, c(3:6)] <- lapply(dat[, c(3:6)], undollar)
  as_tibble(dat)
}



R/contract-convert.R
#' Convert contract names to factor intervals
#'
#' Performs one of three conversions in [market_price()] and [market_history()]:
#' 1. For interval contracts (e.g., "220 - 229", "9% or more", etc.), convert
#' the character strings to proper interval notation.
#' 2. For contracts with multiple discrete outcomes (e.g., Candidate names),
#' convert the character vector to simple factors.
#' 3. For markets with a single binary question (e.g., "Will the Democrats have
#' a brokered convention in 2020?"), contracts returned are always "Yes" which
#' is converted to `TRUE`.
#'
#' @param x A character vector of contract names.
#' @param decimal Should percentages be converted to decimals?
#' @return A interval factor, unique factor, or logical vector.
#' @examples
#' contract_convert(c("4 or fewer", "8 or more", "5", "6", "7"))
#' contract_convert(c("3%-", "12%+", "9% - 12%", "3% - 6%", "6% - 9%"))
#' contract_convert(c("219-", "220 to 229", "240+",  "230 to 239"))
#' contract_convert(c("Sanders", "Biden", "Warren", "Buttigieg"))
#' contract_convert(c("Brokered convention?", "Brokered convention?"))
#' @importFrom stringr str_remove_all str_replace str_replace_all str_detect str_extract_all
#' @export
contract_convert <- function(x, decimal = FALSE) {
  rx <- "(\\d{1,3}(?:,\\d{3})*(?:\\.\\d+)?)"
  p <- all(stringr::str_detect(x, "%"))
  if (length(unique(x)) == 1) {
    return(TRUE)
  } else if (all(stringr::str_detect(x, "\\d", negate = TRUE))) {
    return(factor(x, ordered = FALSE))
  }
  x <- stringr::str_replace(x, "\\sor\\s(fewer|lower)", "-")
  x <- stringr::str_replace(x, "\\sor\\s(more|higher)", "+")
  x <- stringr::str_replace(x, "\\sto\\s", " - ")
  x <- stringr::str_remove_all(x, "[^[\\+\\d\\s\\.-]]")
  x <- stringr::str_remove(x, "(?<=\\d)(\\.0)(?=\\D)")
  if (any(stringr::str_detect(x, sprintf("^%s$", rx)))) {
    x <- stringr::str_replace(x, paste0(rx, "(?:-$)"), "[0,\\1]")
    x <- stringr::str_replace(x, paste0(rx, "(?:\\+$)"), "[\\1, Inf)")
    x <- stringr::str_replace(x, sprintf("^%s$", rx), "[\\1,\\1]")
  } else if (p) {
    x <- stringr::str_replace(x, paste0(rx, "(?:-$)"), "[0,\\1)")
    x <- stringr::str_replace(x, paste0(rx, "(?:\\+$)"), "[\\1,100]")
    x <- stringr::str_replace(x, paste(rx, "-", rx), "[\\1,\\2)")
    if (decimal) {
      x <- stringr::str_replace_all(x, rx, function(p) as.numeric(p)/100)
    }
    n <- stringr::str_extract_all(x, "(\\d{1,3}(?:\\.\\d+)?)|Inf")
    if (length(n) >= 2) {
      d <- abs(diff(as.numeric(c(n[[2]][1], n[[1]][2]))))
      if (p & round(d, digits = 2) == 0.1) {
        x <- stringr::str_replace_all(
          string = x,
          pattern = paste0(rx, "(?=\\)$)"),
          replacement = function(n) as.numeric(n) + 0.1
        )
      }
    }
  } else if (any(stringr::str_detect(x, paste(rx, "-", rx)))) {
    x <- stringr::str_replace(x, paste0(rx, "(?:-$)"), "[0,\\1]")
    x <- stringr::str_replace(x, paste(rx, "(?:-|to)", rx), "[\\1,\\2]")
    x <- stringr::str_replace(x, paste0(rx, "(?:\\+$)"), "[\\1, Inf)")
  }
  z <- unique(x)
  a <- stringr::str_extract_all(z, "(\\d{1,3}(?:\\.\\d+)?)|Inf")
  a <- vapply(a, FUN = function(i) sum(as.numeric(i)), FUN.VALUE = double(1))
  a <- match(a, sort(a))
  factor(x, levels = z[order(sort(z)[a])], ordered = TRUE)
}


R/deprec-contract-convert.R
#' Convert contract names to factor intervals
#'
#' Can perform one of three **rough** conversions:
#' 1. For interval contracts (e.g., "220 - 229", "9% or more", etc.), convert
#' the character strings to proper interval notation.
#' 2. For contracts with multiple discrete outcomes (e.g., Candidate names),
#' convert the character vector to simple factors.
#' 3. For markets with a single binary question (e.g., "Will the Democrats have
#' a brokered convention in 2020?"), contracts returned are always "Yes" which
#' is converted to `TRUE`.
#'
#' @param x A character vector of contract names.
#' @param decimal Should percentages be converted to decimals?
#' @return A interval factor, unique factor, or logical vector.
#' @importFrom stringr str_remove_all str_replace str_replace_all str_detect str_extract_all
# contract_convert(c("4 or fewer", "8 or more", "5", "6", "7"))
# contract_convert(c("3%-", "12%+", "9% - 12%", "3% - 6%", "6% - 9%"))
# contract_convert(c("219-", "220 to 229", "240+",  "230 to 239"))
# contract_convert(c("Sanders", "Biden", "Warren", "Buttigieg"))
# contract_convert(c("Brokered convention?", "Brokered convention?"))
contract_convert <- function(x, decimal = FALSE) {
  rx <- "(\\d{1,3}(?:,\\d{3})*(?:\\.\\d+)?)"
  p <- all(stringr::str_detect(x, "%"))
  if (length(unique(x)) == 1) {
    return(TRUE)
  } else if (all(stringr::str_detect(x, "\\d", negate = TRUE))) {
    return(factor(x, ordered = FALSE))
  }
  x <- stringr::str_replace(x, "\\sor\\s(fewer|lower)", "-")
  x <- stringr::str_replace(x, "\\sor\\s(more|higher)", "+")
  x <- stringr::str_replace(x, "\\sto\\s", " - ")
  x <- stringr::str_remove_all(x, "[^[\\+\\d\\s\\.-]]")
  x <- stringr::str_remove(x, "(?<=\\d)(\\.0)(?=\\D)")
  if (any(stringr::str_detect(x, sprintf("^%s$", rx)))) {
    x <- stringr::str_replace(x, paste0(rx, "(?:-$)"), "[0,\\1]")
    x <- stringr::str_replace(x, paste0(rx, "(?:\\+$)"), "[\\1, Inf)")
    x <- stringr::str_replace(x, sprintf("^%s$", rx), "[\\1,\\1]")
  } else if (p) {
    x <- stringr::str_replace(x, paste0(rx, "(?:-$)"), "[0,\\1)")
    x <- stringr::str_replace(x, paste0(rx, "(?:\\+$)"), "[\\1,100]")
    x <- stringr::str_replace(x, paste(rx, "-", rx), "[\\1,\\2)")
    if (decimal) {
      x <- stringr::str_replace_all(x, rx, function(p) as.numeric(p)/100)
    }
    n <- stringr::str_extract_all(x, "(\\d{1,3}(?:\\.\\d+)?)|Inf")
    if (length(n) >= 2) {
      d <- abs(diff(as.numeric(c(n[[2]][1], n[[1]][2]))))
      if (p & round(d, digits = 2) == 0.1) {
        x <- stringr::str_replace_all(
          string = x,
          pattern = paste0(rx, "(?=\\)$)"),
          replacement = function(n) as.numeric(n) + 0.1
        )
      }
    }
  } else if (any(stringr::str_detect(x, paste(rx, "-", rx)))) {
    x <- stringr::str_replace(x, paste0(rx, "(?:-$)"), "[0,\\1]")
    x <- stringr::str_replace(x, paste(rx, "(?:-|to)", rx), "[\\1,\\2]")
    x <- stringr::str_replace(x, paste0(rx, "(?:\\+$)"), "[\\1, Inf)")
  }
  z <- unique(x)
  a <- stringr::str_extract_all(z, "(\\d{1,3}(?:\\.\\d+)?)|Inf")
  a <- vapply(a, FUN = function(i) sum(as.numeric(i)), FUN.VALUE = double(1))
  a <- match(a, sort(a))
  factor(x, levels = z[order(sort(z)[a])], ordered = TRUE)
}


R/market-plot.R
#' Plot market types
#'
#' @param data A data frame created by either [market_history()] or
#'   [market_price()]. For price history plots, [ggplot2::geom_line()] is used;
#'   for current prices, [ggplot2::geom_col()] is used.
#' @param color The number of current top contracts lines highlighted if `data`
#'   is daily or hourly.
#' @return A ggplot object.
#' @examples
#' market_plot(market_history(3633), color = 4)
#' market_plot(market_price(6405))
#' @importFrom ggplot2 ggplot aes_string geom_col scale_fill_viridis_d
#'   scale_y_continuous theme element_blank geom_line scale_color_brewer
#' @importFrom scales dollar
#' @export
market_plot <- function(data, color = 4) {
  bar <- length(unique(data$time)) == 1
  if (bar) {
    ggplot2::ggplot(data, ggplot2::aes_string(x = "contract")) +
      ggplot2::geom_col(ggplot2::aes_string(y = "close", fill = "close")) +
      ggplot2::scale_fill_viridis_c(end = 0.5, guide = FALSE) +
      ggplot2::scale_y_continuous(labels = scales::dollar) +
      ggplot2::theme(panel.grid.major.x = ggplot2::element_blank()) +
      ggplot2::labs(
        title = unique(data$market),
        caption = paste("Source: PredictIt", unique(data$mid)),
        x = "Contract",
        y = "Closing Price"
      )
  } else {
    td <- data$time == (Sys.Date() - 1)
    con_td <- data$contract[td]
    top_close <- sort(data$close[td], decreasing = TRUE)[1:color]
    top_con <- data$contract[td][data$close[td] %in% top_close]
    data_color <- data[data$contract %in% top_con, ]
    ggplot2::ggplot(data_color, ggplot2::aes_string(x = "time", y = "close")) +
      ggplot2::geom_line(data = data, ggplot2::aes_string(group = "contract"), alpha = 0.5) +
      ggplot2::geom_line(ggplot2::aes_string(color = "contract"), size = 2) +
      ggplot2::scale_color_brewer(palette = "Dark2") +
      ggplot2::scale_y_continuous(labels = scales::dollar) +
      ggplot2::theme(legend.position = "bottom") +
      ggplot2::labs(
        title = unique(data$market),
        caption = paste("Source: PredictIt", unique(data$mid)),
        color = "Contract",
        x = "Contract",
        y = "Closing Price"
      )
  }
}
